### PR TITLE
Update gleam, and use a dynamic GL function table. 

### PR DESCRIPTION
**Pull request by emilio**
cc @sotaroikeda

### CODE REVIEW

**Comment by emilio** [comment]
Blocked on https://github.com/servo/gleam/pull/106

**Comment by kvark** [comment]
So `gl::Gl` is a trait, and thus the new code works with the trait object, right?
Is there any benefit of using the unsafe raw pointers versus just `Rc`/`Weak`?

**Comment by emilio** [comment]
@kvark That's right. It avoids some refcount churn. In this case it seemed easy enough to do it, but I'd be fine also with Rc. It may make sense to share gl function tables across threads though, that's something that Rc would prevent, but we don't use that at all I think so it may be ok.

**Comment by emilio** [comment]
Oh, and I hadn't been following the gleam PR btw, the generic proposal looks ok to me. It's kind of sad to use dynamic dispatch, but I was taking that for granted already. Couldn't we make `Gl` a static table and add convenience methods to it instead? I did something similar for `clang-sys` for bindgen (see the functions struct in https://github.com/KyleMayes/clang-sys/pull/45). It'd be a matter of adding all the glue that would be inlined and would just look like:

```rust
#[inline]
fn flush(&self) {
    (self.functions.expect("glFlush not loaded"))()
}
```

**Comment by sotaroikeda** [comment]
>  Couldn't we make Gl a static table and add convenience methods to it instead?

It might be better for performance. But  I want to handle it as another issue, since we want to enable Graphics branch soon.

**Comment by kvark** [comment]
@emilio 
> It avoids some refcount churn.

The extra refcount work would only apply to resource creation/deletion, right? I wouldn't be too concerned and preferred the code clarity (and less unsafety) instead.

> It may make sense to share gl function tables across threads though

Could you explain why we'd need that? It sounds like a hack.

> Couldn't we make Gl a static table and add convenience methods to it instead?

Possible, although I'd rather want to see static dispatch here in the first place. Considering the deadlines, this will have to wait.

**Comment by emilio** [comment]
On Wed, Feb 15, 2017 at 09:43:07PM -0800, Dzmitry Malyshau wrote:
> @emilio 
> > It avoids some refcount churn.
> 
> The extra refcount work would only apply to resource creation/deletion, right? I wouldn't be too concerned and preferred the code clarity (and less unsafety) instead.

Right, and that's a fair point.

> > It may make sense to share gl function tables across threads though
> 
> Could you explain why we'd need that? It sounds like a hack.

Didn't have a specific use case in mind, but yeah, I guess context
creation time is not a big deal.

I could imagine us sharing the same function hooks or something, but
yeah, as I said I don't expect it to be super-useful.

> > Couldn't we make Gl a static table and add convenience methods to it instead?
> 
> Possible, although I'd rather want to see static dispatch here in the
> first place. Considering the deadlines, this will have to wait.

That's fair, as I said I don't love dynamic dispatch, but we can
probably live with it.

**Comment by kvark** [comment]
@emilio thanks!
I've just published gleam-0.3.0, so let's have this PR updated.

**Comment by sotaroikeda** [comment]
@emilio, thank you!

**Comment by emilio** [comment]
Done, feel free to take a look, let me know when do you want to merge this.

File: src/gl_context.rs
Comment (kvark) [inline src/gl_context.rs]: it appears that we need to sort this out before merging

File: src/draw_buffer.rs
Comment (kvark) [inline src/draw_buffer.rs:41]: why do we need to `destroy` now instead of `drop`?

**Review by kvark — CHANGES_REQUESTED** [review]

File: src/gl_context.rs
Comment (emilio) [reply src/gl_context.rs]: Oh, right, I just grabbed that from the original Sotaro's patch. I guess the idea here would be doing so per-backend (that is, the egl back-end would load gles functions, etc). That being said, in which case would it make a difference? I guess we can provide only the GLES-like interface given this is primarily for WebGL and we create compat contexts, alternatively, I can make it generic, but that cries to avoid the virtual dispatch then.

File: src/draw_buffer.rs
Comment (emilio) [reply src/draw_buffer.rs:41]: Didn't want to grow them by an extra pointer given we only create one per draw buffer, but I can definitely do it I guess.

Comment (kvark) [reply src/draw_buffer.rs:41]: Ah, you are passing `gl` in, I see. And the function is not public, so there is no concern about calling it twice.
Let's leave it with `destroy` then :)

File: src/gl_context.rs
Comment (kvark) [inline src/gl_context.rs:81]: I'm a bit confused about this. `create_shared_with_dispatcher` calls `new_shared_with_dispatcher`, and `new_shared_with_dispatcher` calls `create_shared_with_dispatcher`, both seem unconditional.

File: src/gl_context.rs
Comment (kvark) [reply src/gl_context.rs]: > alternatively, I can make it generic, but that cries to avoid the virtual dispatch then

Well, doing static dispatch is still on the table, just for a different time frame - https://github.com/servo/gleam/issues/107
So maybe it's not the worst idea to be ready for that ;)

Really though, If this newly created context is shared with some existing one, we should use Gl function loader if that existing one is GL, and GLES correspondingly. If it's an entirely new one (`shared_with.is_none() == true`), this is probably where we want the `gleam::GlType` flag to be passed in for run-time context type selection.

**Review by kvark — CHANGES_REQUESTED** [review]

File: src/gl_context.rs
Comment (emilio) [reply src/gl_context.rs:81]: It... doesn't? `create_shared_with_dispatcher` calls the relevant method on the underlying platform-dependent context.

File: src/gl_context.rs
Comment (emilio) [reply src/gl_context.rs]: That sounds reasonable, though probably for `WebGL` it's enough with loading unconditionally just the gles functions.

Comment (emilio) [reply src/gl_context.rs]: tl;dr; I'll prepare a patch with that.

**Comment by sotaroikeda** [comment]
What is a current status?

**Comment by emilio** [comment]
Whoops, totally forgot about this, there it is.

File: src/draw_buffer.rs
Comment (kvark) [inline src/draw_buffer.rs]: `debug_assert_eq!` would be much better here since it would print the actual error/status

File: src/draw_buffer.rs
Comment (kvark) [inline src/draw_buffer.rs]: perhaps, there is no longer a point in this comment? There is no `unsafe` here, so safety is assumed by default

File: src/gl_context.rs
Comment (kvark) [reply src/gl_context.rs:81]: I see now, you are correct.

File: src/tests.rs
Comment (kvark) [inline src/tests.rs]: could use `default()`, which is already implemented on `GlType`

**Review by kvark — APPROVED** [review]
Thanks @emilio , looks good!
I got a few minor things, which are fine to address separately, if you wish.

**Comment by emilio** [comment]
Thanks for the review @kvark! Landed and published as 0.7.0

**Comment by sotaroikeda** [comment]
Thank you!

**Pull request merged**
